File: services/ingestion-openmeteo/app/Application/Outbox/Services/OutboxMessageFactory.php
<?php

namespace App\Application\Outbox\Services;

use App\Application\Outbox\DTO\OutboxMessageData;
use App\Domain\Shared\Event\DomainEvent;
use App\Support\Messaging\RoutingKeyResolver;

/**
 * Ova klasa prevodi domain evente u outbox poruke koje će kasnije
 * publisher poslati na RabbitMQ.
 *
 * Time jasno razdvajamo:
 * - domain event (interna domen logika)
 * - outbox zapis (persistirani integration-ready zapis)
 */
final readonly class OutboxMessageFactory
{
    public function __construct(private RoutingKeyResolver $routingKeyResolver) {}

    /**
     * Kreira jednu outbox poruku iz domain eventa.
     */
    public function fromDomainEvent(DomainEvent $event, string $producer): OutboxMessageData
    {
        return new OutboxMessageData(
            eventId: $event->eventId(),
            eventName: $event->eventName(),
            eventVersion: $event->eventVersion(),
            routingKey: $this->routingKeyResolver->resolve($event->eventName()),
            deduplicationKey: $this->deduplicationKey($event),
            payload: [
                'event_id' => $event->eventId(),
                'event_name' => $event->eventName(),
                'event_version' => $event->eventVersion(),
                'occurred_at' => $event->occurredAt(),
                'producer' => $producer,
                'payload' => $event->payload(),
            ],
            headers: [
                'x-event-id' => $event->eventId(),
                'x-event-name' => $event->eventName(),
                'x-event-version' => $event->eventVersion(),
                'x-producer' => $producer,
            ],
            status: 'pending',
            attempts: 0,
            availableAt: now()->toIso8601String(),
        );
    }

    /**
     * Kreira stabilan key za latest-state outbox zapis.
     */
    private function deduplicationKey(DomainEvent $event): string
    {
        $payload = $event->payload();

        return implode(':', [
            $event->eventName(),
            (string) ($payload['location_id'] ?? ''),
            (string) ($payload['source'] ?? ''),
        ]);
    }

    /**
     * Kreira više outbox poruka iz niza domain eventa.
     *
     * @param DomainEvent[] $events
     * @return OutboxMessageData[]
     */
    public function fromDomainEvents(array $events, string $producer): array
    {
        return array_map(
            fn (DomainEvent $event) => $this->fromDomainEvent($event, $producer),
            $events
        );
    }
}

File: services/ingestion-weatherapi/internal/application/outbox/services/outbox_message_factory.go
package services

import (
	"time"

	outboxdto "ingestion-weatherapi/internal/application/outbox/dto"
	sharedEvent "ingestion-weatherapi/internal/domain/shared/event"
)

/*
OutboxMessageFactory prevodi domain evente u canonical outbox poruke.

Time jasno razdvajamo:
- domain event koji nastaje u agregatu
- outbox zapis koji je spreman za persistence i kasniji publish
*/
type OutboxMessageFactory struct {
	serviceName string
}

/*
NewOutboxMessageFactory pravi novu instancu factory-ja.
*/
func NewOutboxMessageFactory(serviceName string) *OutboxMessageFactory {
	return &OutboxMessageFactory{
		serviceName: serviceName,
	}
}

/*
FromDomainEvent prevodi jedan domain event u jednu outbox poruku.
*/
func (f *OutboxMessageFactory) FromDomainEvent(event sharedEvent.DomainEvent) outboxdto.OutboxMessageData {
	return outboxdto.OutboxMessageData{
		EventID:          event.EventID(),
		EventName:        event.EventName(),
		EventVersion:     event.EventVersion(),
		RoutingKey:       event.EventName(),
		DeduplicationKey: f.deduplicationKey(event),
		Payload: map[string]any{
			"event_id":      event.EventID(),
			"event_name":    event.EventName(),
			"event_version": event.EventVersion(),
			"occurred_at":   event.OccurredAt(),
			"producer":      f.serviceName,
			"payload":       event.Payload(),
		},
		Headers: map[string]any{
			"x-event-id":      event.EventID(),
			"x-event-name":    event.EventName(),
			"x-event-version": event.EventVersion(),
			"x-producer":      f.serviceName,
		},
		Status:      "pending",
		Attempts:    0,
		AvailableAt: time.Now().UTC().Format(time.RFC3339),
		PublishedAt: nil,
		LastError:   nil,
	}
}

/*
deduplicationKey kreira stabilan key za latest-state outbox zapis.
*/
func (f *OutboxMessageFactory) deduplicationKey(event sharedEvent.DomainEvent) string {
	payload := event.Payload()

	locationID, _ := payload["location_id"].(string)
	source, _ := payload["source"].(string)

	return event.EventName() + ":" + locationID + ":" + source
}

/*
FromDomainEvents prevodi više domain eventa u outbox poruke.
*/
func (f *OutboxMessageFactory) FromDomainEvents(events []sharedEvent.DomainEvent) []outboxdto.OutboxMessageData {
	messages := make([]outboxdto.OutboxMessageData, 0, len(events))

	for _, event := range events {
		messages = append(messages, f.FromDomainEvent(event))
	}

	return messages
}
